### PR TITLE
Add ability to encrypt and decrypt all UTF-8 strings.

### LINKED ISSUE #11: Encryption does not work for UTF-8 Strings.
Julia Strings are by default UTF-8 encoded and thus can't simply be used like arrays of UInt8. The code currently assumes that any character in a given String can be converted to UInt8 without a problem, but this doesn't work:

```
julia> str = "UTF-8! 😉"
"UTF-8! 😉"

julia> cipher = AESCipher();

julia> encrypt(str, cipher)
ERROR: InexactError: trunc(UInt8, 128521)
 [...]
```

I've forked this project and I'm trying to fix this properly, but that has exposed some other problems as well - I'll try to make a PR if I get this and the other problems exposed by this fixed.

### PR DESCRIPTION
**Pull request by Seelengrab**
Fixes #11 .

I've also changed some other code to make this change possible and extensible. `decrypt` now always returns `AbstractArray{UInt8}`, since guessing whether or not the original text was valid UTF-8 seemed like a bad idea. The user should handle this. 

Additionally, I've cleaned up some of the code base to make it more idiomatic, but this was by no means comprehensive.

### CODE REVIEW

**Comment by Seelengrab** [comment]
I also think there's possibly some other unrelated issues regarding generic key lengths, but I haven't investigated those. There should probably be more tests in general for other edge cases and special problems.

Additionally, once a `Cipher` has been used to encrypt data, it could maybe be finalized to no longer allow encrypting new data with the same object. This would force e.g. CTR mode ciphers to be recreated and prevent reuse.

File: src/types.jl [hunk 1]
```diff
@@ -59,26 +50,22 @@ function gen_cache(key::AbstractAESKey, mode=CBC)
...
-"""
-mutable struct AESCipher{mode,cacheType,keyType} <: AbstractCipher
+mutable struct Cipher{mode,cacheType,keyType} <: AbstractCipher
 	cache::cacheType
 	key::keyType
 end
+const AESCipher = Cipher
 
-function AESCipher(;key_length=128,mode=CBC,key=keygen(key_length))
+function Cipher(;key_length=128,mode=CBC,key=keygen(key_length))
```
Comment (kanav99) [inline src/types.jl:68]: Don't change the type names, Cipher is too generic

File: src/decrypt.jl [hunk 2]
```diff
@@ -1,38 +1,12 @@
...
-		error("Mismatching keylength")
-	end
-	if iscbc(cipher)
-		AESCBC_D!(plaintext, ciphertext.data, ciphertext.iv, cipher.key, cipher.cache; remove_pad=remove_pad)
-	elseif isctr(cipher)
-		AESCTR_D!(plaintext, ciphertext.data, ciphertext.iv, cipher.key, cipher.cache)
-	elseif isecb(cipher)
-		AESECB_D!(plaintext, ciphertext.data, cipher.key, cipher.cache)
-	end
-	return (ciphertext.original_type)(plaintext)
```
Comment (kanav99) [inline src/decrypt.jl:36]: I differ from the opinion that only ByteArray should be returned. But we can do one thing, we should probably make some kwarg like `to_original_type` which should be true by default. If true, make the type cast, otherwise return byte array

File: src/encrypt.jl [hunk 3]
```diff
@@ -1,26 +1,20 @@
...
+
+encrypt!(ciphertext, plaintext, cipher::Cipher{CBC,T,U}; iv=rand(UInt8, 16)) where {T,U} = encrypt!(ciphertext, transcode(UInt8, plaintext), cipher, iv)
+encrypt!(ciphertext, plaintext, cipher::Cipher{CTR,T,U}; iv=rand(UInt8, 16)) where {T,U} = encrypt!(ciphertext, transcode(UInt8, plaintext), cipher, iv)
+encrypt!(ciphertext, plaintext, cipher::Cipher) = encrypt!(ciphertext, transcode(UInt8, plaintext), cipher)
+
+encrypt!(ciphertext, plaintext::AbstractArray{UInt8}, cipher::Cipher{CBC,T,U}, iv) where {T,U} = CipherText(AESCBC!(ciphertext, plaintext, iv, cipher.key, cipher.cache), iv, get_key_length(cipher), CBC)
+encrypt!(ciphertext, plaintext::AbstractArray{UInt8}, cipher::Cipher{CTR,T,U}, iv) where {T,U} = CipherText(AESCTR!(ciphertext, plaintext, iv, cipher.key, cipher.cache), iv, get_key_length(cipher), CTR)
+encrypt!(ciphertext, plaintext::AbstractArray{UInt8}, cipher::Cipher{ECB,T,U}) where {T,U} = CipherText(AESECB!(ciphertext, plaintext, cipher.key, cipher.cache), nothing, get_key_length(cipher), ECB)
```
Comment (kanav99) [inline src/encrypt.jl:20]: I like these changes!

File: src/types.jl (hunk 1 above)
Comment (Seelengrab) [reply src/types.jl:68]: I've changed the type name internally only. Below the definition is `const AESCipher = Cipher` which is exported, so the API doesn't change. Internally, it's clear that everything is an AESCipher since we're in the AES package.

Comment (kanav99) [reply src/types.jl:68]: Ohh makes sense

File: src/decrypt.jl (hunk 2 above)
Comment (Seelengrab) [reply src/decrypt.jl:36]: Well, the problem with returning something other than a vector of bytes is that simply from the ciphertext it's generally impossible to know whether the resulting plaintext is valid UTF-8 or not. Imagine someone trying to decrypt some other text, doesn't know that all Strings in julia must be valid UTF-8 if created via `String` and says "well I want text". That will not work and the decryption itself failed and discards the result, instead of the data still being available to them.

Additionally, having that conversion as a field of the ciphertext leads to type instability, since there _has to be_ a dynamic lookup of the result, which would be good to avoid.

Comment (kanav99) [reply src/decrypt.jl:36]: Yeah that's a valid point, Type instability might creep in. I think still we should find some good solution to it, but as of now this one looks good to merge. We can make it a new issue.

**Comment by kanav99** [comment]
Are you planning more changes in the PR or should this be merged?

**Comment by Seelengrab** [comment]
You can merge it as-is, the other changes I'm thinking about should be a different PR. This PR and the related changes are done :)

**Comment by kanav99** [comment]
Awesome! Thank you!
